### PR TITLE
[consensus] handle messages from future heights

### PR DESCRIPTION
**Pull request by olegrok**
Initial research focused on fixing archive_node test failures. The test was crashing during block generation.
First attempt failed—separating parallel tests by protocol did not help. Further research showed the test crashes even without parallel execution, just less frequently.

This patch fixes the root cause: consensus ignored messages from block N+1 when block N was not yet committed. Now, we use the last known block and take validators from it if needed. Future improvements may split validators by periods, ensuring we know their validity timeframe.

### CODE REVIEW

**Comment by knazarov** [comment]
> This patch fixes the root cause: consensus ignored messages from block N+1 when block N was not yet committed.

I'm not sure how this is possible.Can you give an example?

**Comment by olegrok** [comment]
> > This patch fixes the root cause: consensus ignored messages from block N+1 when block N was not yet committed.
> 
> I'm not sure how this is possible.Can you give an example?

Just a part of logs of failed test:
```
// First instance committed block 256 and start consensus on block 267
2025-02-26 06:49:12 INF [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:375 > sequence done height=[256]	 shardId=[0]	
2025-02-26 06:49:12 INF [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:316 > sequence started height=[257]	 shardId=[0]	
2025-02-26 06:49:12 INF [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:323 > round started round=[0]	 shardId=[0]	
2025-02-26 06:49:12 INF [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:399 > we are the proposer shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:409 > block proposal accepted shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:912 > exit: commit state shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:959 > enter: fin state shardId=[0]	
2025-02-26 06:49:12 DBG [sync]	 github.com/NilFoundation/nil/nil/internal/collate/syncer.go:196 > Received block blockNumber=[256]	 shardId=[0]	
2025-02-26 06:49:12 DBG [validator]	 github.com/NilFoundation/nil/nil/internal/collate/proposer.go:431 > Collected 0 incoming transactions from neigbors with 0 gas shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:963 > exit: fin state shardId=[0]	
2025-02-26 06:49:12 DBG [validator]	 github.com/NilFoundation/nil/nil/internal/collate/proposer.go:122 > Collected 0 internal, 0 external (0 gas) and 0 forward transactions shardId=[0]	
2025-02-26 06:49:12 DBG [sync]	 github.com/NilFoundation/nil/nil/internal/collate/syncer.go:196 > Received block blockNumber=[256]	 shardId=[0]	

// Proposal for block 257 is multicasted
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:413 > pre-prepare message multicasted shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/internal/consensus/ibft/transport.go:90 > Validator message received addr=[36ebfe7335a1b89a5ee969c1940cc78c399381a4d5100e40148e8ea841207d6835c7aa7f62c28f67b0f3ad86a851eaabd882dee6931aaa564db6e923ebf5f23077bcfea2a306ca7d03640bd8b5371fbac218913e3138690941b72f1c61da49e107e6a532d5ee448de4a992b31e781f38af00c3874e96cfef0d03bf816d38c22e]	 height=[257]	 round=[0]	 shardId=[0]	 topic=[/ibft/0.2/shard/0]	 type=[PREPREPARE]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:805 > enter: prepare state shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/internal/consensus/ibft/transport.go:90 > Validator message received addr=[36ebfe7335a1b89a5ee969c1940cc78c399381a4d5100e40148e8ea841207d6835c7aa7f62c28f67b0f3ad86a851eaabd882dee6931aaa564db6e923ebf5f23077bcfea2a306ca7d03640bd8b5371fbac218913e3138690941b72f1c61da49e107e6a532d5ee448de4a992b31e781f38af00c3874e96cfef0d03bf816d38c22e]	 height=[257]	 round=[0]	 shardId=[0]	 topic=[/ibft/0.2/shard/0]	 type=[PREPREPARE]	

// Second node accepted proposal message
// But proposal itself was rejected since we have not yet committed block 256
2025-02-26 06:49:12 ERR [consensus]	 github.com/NilFoundation/nil/nil/internal/consensus/ibft/verifier.go:41 > Failed to get validators error="key not found in db" height=[257]	 publicKey=[36ebfe7335a1b89a5ee969c1940cc78c399381a4d5100e40148e8ea841207d6835c7aa7f62c28f67b0f3ad86a851eaabd882dee6931aaa564db6e923ebf5f23077bcfea2a306ca7d03640bd8b5371fbac218913e3138690941b72f1c61da49e107e6a532d5ee448de4a992b31e781f38af00c3874e96cfef0d03bf816d38c22e]	 round=[0]	 shardId=[0]	
2025-02-26 06:49:12 ERR [consensus]	 github.com/NilFoundation/nil/nil/internal/consensus/ibft/ibft.go:177 > Failed to insert proposal error="received old block" height=[256]	 round=[0]	 shardId=[0]	
2025-02-26 06:49:12 INF [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:375 > sequence done height=[256]	 shardId=[0]	
2025-02-26 06:49:12 DBG [collator]	 github.com/NilFoundation/nil/nil/internal/collate/scheduler.go:134 > Consensus interrupted by syncer shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:409 > block proposal accepted shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:413 > pre-prepare message multicasted shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/internal/consensus/ibft/transport.go:90 > Validator message received addr=[5ad7e33d193f89a90fa6bc35c9a9ffd4dce33914c11b3af37333151a9ce9f77c557e20325a4a0ac49976835ebed6b1c2e1d9ce8de5a8f2c7a2b7f43c84d616474d2f89d42e13fabd80aaa5c18b7b7d969cea23aa08d6ee9e5cfdea6da25a97cc6b41a2ed4d90c59b64ae5abb728232a76f08d575bd65a56db84c7aa61d3a79da]	 height=[257]	 round=[0]	 shardId=[0]	 topic=[/ibft/0.2/shard/0]	 type=[PREPREPARE]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:805 > enter: prepare state shardId=[0]	
2025-02-26 06:49:12 DBG [consensus]	 github.com/NilFoundation/nil/nil/internal/consensus/ibft/transport.go:90 > Validator message received addr=[5ad7e33d193f89a90fa6bc35c9a9ffd4dce33914c11b3af37333151a9ce9f77c557e20325a4a0ac49976835ebed6b1c2e1d9ce8de5a8f2c7a2b7f43c84d616474d2f89d42e13fabd80aaa5c18b7b7d969cea23aa08d6ee9e5cfdea6da25a97cc6b41a2ed4d90c59b64ae5abb728232a76f08d575bd65a56db84c7aa61d3a79da]	 height=[257]	 round=[0]	 shardId=[0]	 topic=[/ibft/0.2/shard/0]	 type=[PREPREPARE]	

// Anyway the second node started consensus on block 257
// But proposal was lost
2025-02-26 06:49:12 INF [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:316 > sequence started height=[257]	 shardId=[0]	
2025-02-26 06:49:12 INF [consensus]	 github.com/NilFoundation/nil/nil/go-ibft/core/ibft.go:323 > round started round=[0]	 shardId=[0]	
```

After that consensus stucks somewhy.

I'm pretty sure that case a bit artificial because in test we have only two validators. But anyway it's better to fix such corner case because it can shoot on our dev stand where we don't have a lot of validators.

File: nil/internal/collate/scheduler.go [hunk 1]
```diff
@@ -122,19 +122,29 @@ func (s *Scheduler) doCollate(ctx context.Context) error {
...
-		case err := <-consCh:
-			return err
+		for {
+			select {
+			case newBlockId := <-syncCh:
+				if newBlockId < types.BlockNumber(height) {
+					continue
+				}
+
+				cancelFn()
```
Comment (knazarov) [inline nil/internal/collate/scheduler.go:138]: Please add a comment here why cancelling consensus here is safe

**Review by knazarov — APPROVED** [review]
